### ISSUE TITLE
Don't bundle CPAN libraries

### ISSUE THREAD

**Issue by phile314** [LMS-Community/slimserver#138]
I am trying to package slimserver for [NixOS](http://nixos.org/). However, I noticed that the repository contains a great amount of copied CPAN libraries. This makes it difficult to package slimserver, as some CPAN packages need patches to work properly on NixOS. Furthermore, using copies also adds a significant security risk as slimserver may be using outdated/vulnerable library versions. 

It would be great if slimserver instead could use CPAN packages directly and not use the bundled versions. I am aware that this probably is a lot of work, but it would be a great help.

I guess this could also help the Arch package @stefansielaff ?

**Comment by mherger** [comment]
Most of the CPAN modules might not be a problem at all: just dump the CPAN sub-folder. You might need to wipe the modules.conf, too.

The other issue is the lib folder: the modules in there have been modified to fix issues or adjust to some need. 

Now that said: it might just work with standard modules. But we can't guarantee. We're aware that the situation isn't great, but it's been working rather fine for years...

**Comment by stefansielaff** [comment]
As a developer (perl is not my predilection) I understand both sides and deal with it. 
This is why I've forked this a while ago and use the fork for the arch packages instead of the original repo. I'm trying to remove as many modules as possible over time, but to be honest this is a brave approach because Arch moves quickly and you can never rely on anything.

**Comment by phile314** [comment]
Just to let you know, just ignoring most of the CPAN modules did work fine. The only two packages that I used from the bundled copies where DBIx and SQL, as it seems that slimserver doesn't work with the current versions of those packages.

If anyone is interested, this are my build specifications:
Package: https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/slimserver/default.nix
Service:   https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/audio/slimserver.nix#L45

**Comment by mherger** [comment]
@phile314 - that's good to know. Thanks!

**Comment by stefansielaff** [comment]
@phile314 I'm late but just read your reply... interesting! 
There are a few more modules in the delivered folder which don't seem to be defined in your script. Starting with Algorithm::C3 - have you simply dropped them or are they pulled in as a dependency on NixOS?

**Comment by phile314** [comment]
If they are transitive dependencies of some other dependency the get pulled in, else not. I just kept adding libraries until things started working. It could be that I am missing some libraries and some functionality might be broken. Haven't noticed any breakage yet though.

**Comment by paulgevers** [comment]
I concur that it seems to be working with SQL and DBIx from slimserver with the rest removed. Without those I get errors like below. I haven't checked yet if it is either SQL or DBIx or both. Either way, it seems modules.conf should be updated with a maximum version to reflect this. The upper limit for Audio::Scan seems unneeded though as 0.98 seems to work.
```
[17-11-30 22:14:20.7100] main::init (387) Starting Logitech Media Server (v7.9.1, 1511690631, Sun Nov 26 10:03:51 UTC 2017) perl 5.026001 - x86_64-linux-gnu-thread-multi
[17-11-30 22:14:22.9816] Slim::Utils::Misc::msg (1250) Warning: [22:14:22.9809] DBIx::Class::Componentised::inject_base(): Use of DBIx::Class::UTF8Columns is strongly discouraged. See documentation of DBIx::Class::UTF8Columns for more info
[17-11-30 22:14:23.7313] Slim::Schema::Storage::throw_exception (122) Error: DBI Exception: DBD::SQLite::db do failed: database persistentdb is already in use
[17-11-30 22:14:23.7330] Slim::Schema::Storage::throw_exception (122) Backtrace:

   frame 0: Slim::Utils::Log::logBacktrace (/usr/share/perl5/Slim/Schema/Storage.pm line 122)
   frame 1: Slim::Schema::Storage::throw_exception (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 1501)
   frame 2: DBIx::Class::Storage::DBI::__ANON__ (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 1468)
   frame 3: DBIx::Class::Storage::DBI::__ANON__ (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 856)
   frame 4: DBIx::Class::Storage::DBI::__ANON__ (/usr/share/perl5/DBIx/Class/Storage/BlockRunner.pm line 130)
   frame 5: DBIx::Class::Storage::BlockRunner::try {...}  (/usr/share/perl5/Try/Tiny.pm line 103)
   frame 6: (eval) (/usr/share/perl5/Try/Tiny.pm line 94)
   frame 7: Try::Tiny::try (/usr/share/perl5/DBIx/Class/Storage/BlockRunner.pm line 134)
   frame 8: DBIx::Class::Storage::BlockRunner::__ANON__ (/usr/share/perl5/Context/Preserve.pm line 23)
   frame 9: Context::Preserve::preserve_context (/usr/share/perl5/DBIx/Class/Storage/BlockRunner.pm line 213)
   frame 10: DBIx::Class::Storage::BlockRunner::_run (/usr/share/perl5/DBIx/Class/Storage/BlockRunner.pm line 105)
   frame 11: DBIx::Class::Storage::BlockRunner::run (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 857)
   frame 12: DBIx::Class::Storage::DBI::dbh_do (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 1470)
   frame 13: DBIx::Class::Storage::DBI::_do_query (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 1426)
   frame 14: DBIx::Class::Storage::DBI::connect_call_do_sql (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 1391)
   frame 15: DBIx::Class::Storage::DBI::try {...}  (/usr/share/perl5/Try/Tiny.pm line 103)
   frame 16: (eval) (/usr/share/perl5/Try/Tiny.pm line 94)
   frame 17: Try::Tiny::try (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 1419)
   frame 18: DBIx::Class::Storage::DBI::_do_connection_actions (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 1401)
   frame 19: DBIx::Class::Storage::DBI::try {...}  (/usr/share/perl5/Try/Tiny.pm line 103)
   frame 20: (eval) (/usr/share/perl5/Try/Tiny.pm line 94)
   frame 21: Try::Tiny::try (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 1419)
   frame 22: DBIx::Class::Storage::DBI::_do_connection_actions (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 1065)
   frame 23: DBIx::Class::Storage::DBI::_run_connection_actions (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 1058)
   frame 24: DBIx::Class::Storage::DBI::_populate_dbh (/usr/share/perl5/DBIx/Class/Storage/DBI.pm line 947)
   frame 25: DBIx::Class::Storage::DBI::ensure_connected (/usr/share/perl5/Slim/Schema/Storage.pm line 43)
   frame 26: (eval) (/usr/share/perl5/Slim/Schema/Storage.pm line 43)
   frame 27: Slim::Schema::Storage::dbh (/usr/share/perl5/Slim/Schema.pm line 287)
   frame 28: Slim::Schema::_connect (/usr/share/perl5/Slim/Schema.pm line 130)
   frame 29: Slim::Schema::init (/usr/sbin/squeezeboxserver line 552)
   frame 30: main::init (/usr/sbin/squeezeboxserver line 675)
   frame 31: main::main (/usr/sbin/squeezeboxserver line 1213)
```

**Comment by fsbruva** [comment]
> I concur that it seems to be working with SQL and DBIx from slimserver with the rest removed. Without those I get errors like below. I haven't checked yet if it is either SQL or DBIx or both.

This is a known issue with DBIx, as documented in [this thread.](http://forums.slimdevices.com/showthread.php?89569-quot-database-persistentdb-is-already-in-use-quot-error-with-7-6)

@mherger  Is the scope to make slimserver detect & use newer DBIx known? Is it simply not worth the effort?

**Comment by mherger** [comment]
I never looked into this. Feel free to give it a try. But if Andy at the time didn't want to move ahead, then I think it might be non-trivial.

**Comment by adamcstephens** [comment]
From my testing, the vast majority of the vendored modules can be replaced with CPAN modules. In nixpkgs we recently bumped perl to 5.38 which broke slimserver, but I've since proven that it can work with 5.38. We'd rather not use the slimserver-vendor repo or built in compiled modules if possible. https://github.com/NixOS/nixpkgs/pull/265556

**Comment by gordonb3** [comment]
Just stumbled across this. I am maintaining a package for Gentoo Linux as part of an ARMv5 based home server operating system. Like you I found that LMS is seriously picky about the DBIx version. I have not detected any issue with SQL. Initially I ran the slimserver-vendor routine to build the arch specific modules but I didn't like it one bit that some of the modules required literally hours to compile as they statically linked to antique versions of e.g. ffmpeg.

I found that most of the required modules for running LMS were also present as system provided packages (referencing CPAN) and in some cases even already installed as part of the OS web based GUI. Which was yet another nuisance as Perl prefers system path for loading libraries but LMS sets the module load path to prefer its own CPAN and Slim folders - which leads to cryptic error messages about versions being incorrect.

I created additional packages for the missing modules from slimserver-vendor, using the system default of dynamic linking to modern support libraries. The result is that my LMS installation does not contain any arch specific modules at all, however this does not eliminate all of the conflicts I mentioned above. To accomplish that my final step in installing LMS is a cyclic removal of every module instance within the CPAN folder that also exists in the Perl system path (except of course DBIx). I have stepped up from 5.34 to 5.36 and 5.38 without needing to make any changes to this routine.

**Comment by adamcstephens** [comment]
The list we're keeping from the modules vendored in this repo are listed here: https://github.com/NixOS/nixpkgs/blob/776dfd7eaab2aa41d7cb1c6c0619e04032fe0535/pkgs/servers/slimserver/default.nix#L114-L125

I did end up using the AudioScan module version from slimserver-vendor, as its newer than CPAN and some users were experiencing problems with the CPAN version. https://github.com/NixOS/nixpkgs/blob/f9d39fb9aff0efee4a3d5f4a6d7c17701d38a1d8/pkgs/top-level/perl-packages.nix#L1295-L1310

**Comment by gordonb3** [comment]
Yes. CPAN is stuck at 1.01 As I didn't want to break the Gentoo package method for Perl modules and thus risk Audio::Scan to stay behind on a Perl upgrade I added a patch to version 1.02 in my repository. I've been adding quite a few more patches the last year...

https://github.com/gordonb3/bubba-overlay/tree/master/dev-perl/Audio-Scan

**Comment by adamcstephens** [comment]
I'd still love to see this resolved at some point, but understand if it's never going to be tackled due to the complexity.

**Comment by paulgevers** [comment]
I like to note that I [started](https://bugs.debian.org/1132555) the work to ship lyrionmusicserver in Debian.

Apart from the notorious DBIx version I'd like to share that I needed to keep the version of `CPAN/File/Next.pm` clamped at the version shipped by `slimserver` as the version in Debian is newer and didn't work.

**Comment by michaelherger** [comment]
How/why did it fail with `File::Next` updated? Any specific error?

**Comment by paulgevers** [comment]
> How/why did it fail with `File::Next` updated? Any specific error?

I'm still learning how to expose failures in the log. I couldn't get the error yet, but the scanning of files didn't find any music file with the Debian version of `File::Next` and did with the vendored version.

**Comment by adamcstephens** [comment]
I went through a bunch of effort de-vendoring as much as possible in the nixpkgs package. You can see what our setup looked like here https://github.com/NixOS/nixpkgs/blob/107cba9eb4a8d8c9f8e9e61266d78d340867913a/pkgs/by-name/sl/slimserver/package.nix 

When perl 5.41 was added to nixpkgs, one or more modules broke. Adding on to a few various issues with de-vendoring over the past couple years, including multiple scan related issues, and I decided to pull the plug on the effort. We now only use nixpkgs perl modules for deps that are compiled native code, and use the vendored versions for everything else. https://github.com/NixOS/nixpkgs/blob/8110df5ad7abf5d4c0f6fb0f8f978390e77f9685/pkgs/by-name/sl/slimserver/package.nix

I'd rather de-vendor everything and use our own packages, but I'm not convinced it's worth the effort unless the LMS project devs are interested in contributing and giving it first class support.

**Comment by gordonb3** [comment]
Sorry about all your time spent. I went through all of this some 6 years ago, as I switched from using the CPAN `buildme.sh` script to system provided packages. I started out with a fairly extensive list of included modules that I thought were safe to delete, and even then I let a post install routine recursively look for modules that were already in the system vendor path and remove those as well from the CPAN (and lib) folder. This worked up until mid 2024 and then some system provided module stopped being compatible with LMS.

So I reviewed my method of deleting as many CPAN files as seemed safe, since obviously it wasn't. I now only use the recursive removal of files for modules that occur in the system vendor arch folder which I find by
`LANG="en_US.UTF-8" LC_ALL="en_US.UTF-8" /usr/bin/perl -V | grep -o "vendorarch[^ ]* "`
Ideally this should also not be needed, but I found that a number of these modules are included in the base CPAN folder rather than just in CPAN/arch. I do understand the economic reasoning behind it though.

One other thing: I noticed in your installer package that you are grabbing LMS from the git repository, which includes an enormous amount of binaries that you have no use for. There is also a download site where you can grab significantly smaller archives for specific platforms, or for our use cases a download that does not include any vendor arch files at all. For the 9.1.0 version this is:
`https://downloads.lms-community.org/LyrionMusicServer_v9.1.0/lyrionmusicserver-9.1.0-noCPAN.tgz`
Do note that this link is case sensitive.

**Comment by adamcstephens** [comment]
> you are grabbing LMS from the git repository

Yes, this is by design. We strongly prefer git sources to release tarballs for security reasons.

**Comment by paulgevers** [comment]
For other distribution packagers: I think I found a solution for the DBIx::Class problem. See #1562. Comments extremely welcome.

**Comment by mavit** [comment]
> I'd rather de-vendor everything and use our own packages, but I'm not convinced it's worth the effort unless the LMS project devs are interested in contributing and giving it first class support.

Realistically, if us packagers want to see this, I think we're going to need to drive the upstream work required between ourselves.

We can hope that we reach a tipping point where it's easier to patch LMS to work with the latest version of a given CPAN module than it is to rebuild a specific old version of that module for the latest version of Perl adopted by distros.  However, that rebuilding needs to be done for non-Linux platforms anyway, so perhaps it's not a big win.

**Comment by gordonb3** [comment]
> Realistically, if us packagers want to see this, I think we're going to need to drive the upstream work required between ourselves.
> 
> We can hope that we reach a tipping point where it's easier to patch LMS to work with the latest version of a given CPAN module than it is to rebuild a specific old version of that module for the latest version of Perl adopted by distros. However, that rebuilding needs to be done for non-Linux platforms anyway, so perhaps it's not a big win.

Honestly, the only packages that requires our concern are the ones that are linked to arch specific `.so` libraries. This is due to how Perl prioritizes search paths, causing the dreaded perl loader version mismatch error. System provided DBIx module being a different version from LMS provided DBIx module does not create a problem, because LMS will prefer the one from its own CPAN folder and there is no `.so` associated with it. There are however a number of modules that are directly part of CPAN, but should be below CPAN/arch as they do reference a `.so` that can also be overridden by a system provided package. If these files were moved to where they should be, and as such not be included in the `no-CPAN` archive, then LMS would run straight out of the box with system provided CPAN/arch matching modules on literally every system.

**Comment by paulgevers** [comment]
> Realistically, if us packagers want to see this, I think we're going to need to drive the upstream work required between ourselves.

I totally agree with this. But it can go in pieces. I mean, if we first get the changes landed in lms that are required to work both with the shipped versions as well as the versions in our distributions, it's already easier to switch for us. And if we show that with those changes lms "just works" with newer libraries, half of the work to update lms itself to ship the newer versions is done.

In Debian, I'm down to 15 files (tree shown at the end of this message), and that even includes two libraries I still need to ship as stand alone packages. And I'll admit that I'm currently fighting with two issues because of the move to the latest version of `DBIx`. I'll not claim everything works as I'm sure my testing only covers a small part of lms, but it's a data point. It's shipped in `experimental` only at the moment because of that.

> We can hope that we reach a tipping point where it's easier to patch LMS to work with the latest version of a given CPAN module than it is to rebuild a specific old version of that module for the latest version of Perl adopted by distros. However, that rebuilding needs to be done for non-Linux platforms anyway, so perhaps it's not a big win.

What I'm seeing with `DBIx` is that there's a bunch of warnings and errors that are probably genuine issues with the current code and are visible because `DBIx` got better over time in checking. While it's work moving forward, I expect there's also improvements coming out of that. We'll see.

```
paul@toba ~/packages/lms/slimserver $ tree CPAN lib
CPAN
├── JSON
│   └── XS
│       └── VersionOneAndTwo.pm
├── MP3
│   └── Cut
│       ├── Gapless
│       │   └── Track.pm
│       └── Gapless.pm
├── Tie
│   ├── Cache
│   │   └── LRU
│   │       └── Expires.pm
│   └── RegexpHash.pm
└── XML
    ├── XSPF
    │   ├── Base.pm
    │   └── Track.pm
    └── XSPF.pm
lib
├── AnyEvent
│   └── Impl
│       └── PerfMonEV.pm
├── Async
│   └── Util.pm
├── DBIx
│   ├── Class
│   │   └── ResultSet.pm
│   └── Migration.pm
├── MPEG
│   └── Audio
│       └── Frame.pm
├── Protocol
│   └── WebSocket
│       └── Frame.pm
└── README

21 directories, 15 files
```

**Comment by mavit** [comment]
I expect some of you folks will have opinions on a bug I just raised, #1640.
